src/metabase/actions/execution.clj

```diff
@@ -172,12 +172,25 @@
 (defn execute-action!
   "Execute the given action with the given parameters of shape `{<parameter-id> <value>}."
   [action request-parameters]
-  (case (:type action)
-    :implicit
-    (execute-implicit-action action request-parameters)
-    (:query :http)
-    (execute-custom-action action request-parameters)
-    (throw (ex-info (tru "Unknown action type {0}." (name (:type action))) action))))
+  (let [; add default values for missing parameters
+        all-param-ids          (set (map :id (:parameters action)))
+        provided-param-ids     (set (keys request-parameters))
+        missing-param-ids      (set/difference all-param-ids provided-param-ids)
+        missing-param-defaults (into {}
```

Comment:
what happens if a param doesn't have value and defaultValue? should we throw an error?

Reply:
Good question. The answer is no error should be thrown (at least here). It's always been ok to not provide a param if it's optional. Previously not providing a param is the same as if the param is present in request-parameters but it's value is `nil`, which happens when a user submits an action with a blank optional form field.

If it's an implicit action, the value will just be ignored and won't be a parameter in the implicit query from `build-implicit-query`. For example for when `kind=row/update`, omitting a parameter will mean the query does not update that parameter.

If it's a query action, then what happens depends on whether the template tag is wrapped in `[[ ]]` or not, following the same logic as template tags in native queries. For example, given this SQL:
```
UPDATE orders 
SET quantity = {{ quantity }}
[[WHERE id = {{ id }}]]
```
If `id` is not provided, this won't result in an error. If `quantity` is not provided, this will result in an error.

I just added a test to this PR [here](https://github.com/metabase/metabase/commit/8ac370863551c5276f4d77aeac56c10ac9346021), because this case wasn't explicitly tested for action execution.

Reply:
that makes sense, thanks for the explanation!